File: src/main/java/org/radarbase/appserver/entity/TaskStateEvent.java
package org.radarbase.appserver.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import org.radarbase.appserver.event.state.TaskState;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.time.Instant;

@Entity
@Getter
@Table(name = "task_state_events")
@NoArgsConstructor
public class TaskStateEvent {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TaskState state;

    @NotNull
    @Column(nullable = false)
    private Instant time;

    @Column(name = "associated_info", length = 1250)
    private String associatedInfo;

    @NotNull
    @ManyToOne(fetch = FetchType.EAGER, optional = false)
    @JoinColumn(name = "task_id", nullable = false)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JsonIgnore
    private Task task;


    public TaskStateEvent(
            @NotNull Task task,
            @NotNull TaskState state,
            @NotNull Instant time,
            String associatedInfo
    ) {
        this.state = state;
        this.time = time;
        this.associatedInfo = associatedInfo;
        this.task = task;
    }
}

File: src/main/java/org/radarbase/appserver/entity/TaskStateEvent.kt
package org.radarbase.appserver.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import jakarta.persistence.*
import jakarta.validation.constraints.NotNull
import org.hibernate.annotations.OnDelete
import org.hibernate.annotations.OnDeleteAction
import org.radarbase.appserver.event.state.TaskState
import java.time.Instant

@Entity
@Table(name = "task_state_events")
class TaskStateEvent {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    val id: Long? = null

    @field:NotNull
    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    var state: TaskState? = null

    @field:NotNull
    @Column(nullable = false)
    var time: Instant? = null

    @Column(name = "associated_info", length = 1250)
    var associatedInfo: String? = null

    @field:NotNull
    @field:JsonIgnore
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(name = "task_id", nullable = false)
    @ManyToOne(fetch = FetchType.EAGER, optional = false)
    var task: Task? = null

    constructor()
    constructor(task: Task?, state: TaskState?, time: Instant?, associatedInfo: String?) {
        this.state = state
        this.time = time
        this.associatedInfo = associatedInfo
        this.task = task
    }


}
